core/store/src/utils/shard_size.rs

```diff
@@ -0,0 +1,142 @@
+    // Find the split key
+    let Some(key) = (match trie.read_lock_memtries() {
+        Some(memtries) => find_split_key(MemTrieIteratorInner::new(&*memtries, trie))?,
+        None => find_split_key(DiskTrieIteratorInner::new(trie))?,
+    }) else {
+        return Ok(None);
+    };
+
+    // Convert the key to account ID by finding the longest prefix that is a valid ID
+    for i in 0..key.len() {
+        let prefix = std::str::from_utf8(&key[..(key.len() - i)]).expect("trie key is non-ASCII");
+        if let Ok(account_id) = AccountId::from_str(prefix) {
+            return Ok(Some(account_id));
+        }
+    }
+    Ok(None)
+}
+
+fn find_split_key<NodePtr, ValueRef, TrieStorage>(
+    trie_storage: TrieStorage,
+) -> Result<Option<Vec<u8>>, StorageError>
+where
+    NodePtr: Copy,
+    TrieStorage: GenericTrieInternalStorage<NodePtr, ValueRef>,
+{
+    let mut key_nibbles = Vec::with_capacity(AccountId::MAX_LEN * 2); // x2 because nibbles are half-bytes
+    let Some(mut current_node_ptr) = trie_storage.get_root() else {
+        return Ok(None);
+    };
+    loop {
+        match trie_storage.get_node(current_node_ptr, AccessOptions::DEFAULT)? {
+            GenericTrieNode::Empty => return Ok(None),
+            GenericTrieNode::Leaf { extension, .. } => {
+                key_nibbles.extend(extension);
+                return Ok(Some(NibbleSlice::nibbles_to_bytes(&key_nibbles)));
+            }
+            GenericTrieNode::Extension { extension, child } => {
+                key_nibbles.extend(extension);
+                current_node_ptr = child;
+            }
+            GenericTrieNode::Branch { children, value } => {
+                let Some(parent_value_ref) = value else { return Ok(None) };
+                let parent_contribution = get_contribution(&trie_storage, parent_value_ref)?;
+                if let Some((idx, child)) =
+                    find_middle_child(&trie_storage, parent_contribution, *children)?
+                {
+                    key_nibbles.push(idx);
+                    current_node_ptr = child;
+                } else {
+                    return Ok(Some(NibbleSlice::nibbles_to_bytes(&key_nibbles)));
+                }
+            }
+        }
+    }
+}
+
+fn find_middle_child<NodePtr, ValueRef, TrieStorage>(
+    trie_storage: &TrieStorage,
+    parent_contribution: ShardSizeContribution,
+    children: [Option<NodePtr>; 16],
+) -> Result<Option<(u8, NodePtr)>, StorageError>
+where
+    NodePtr: Copy,
+    TrieStorage: GenericTrieInternalStorage<NodePtr, ValueRef>,
+{
+    for i in 0..children.len() {
+        let Some(child_ptr) = children[i] else { continue };
+        let Some(child_contribution) = get_node_contribution(trie_storage, child_ptr)? else {
+            continue;
+        };
+        if child_contribution >= parent_contribution / 2 {
+            return Ok(Some((i as u8, child_ptr)));
+        }
```

Comment:
I think you should accumulate the contributions from children until half parent is exceeded and then return either the previous one or the current one.

Reply:
Indeed, it should be accumulated. My mistake.